### ISSUE TITLE
Bug: Following errors gets repeated every time: ERROR: A database consistency issue has been caught. A --resync is needed to rebuild the database

### ISSUE THREAD

**Issue by nathan-gs** [abraunegg/onedrive#3030]
### Describe the bug

During a `resync` following error is shown:
```
ERROR: A database consistency issue has been caught. A --resync is needed to rebuild the database.
```

After running it again with `--resync` the error remains.

I came originally from a 2.4 version, and I am using Onedrive on Windows:
```
OneDrive device ID: 16b6a41c-b0a0-14e2-397f-bb3b4509fa7c
OneDrive version: Build 24.221.1103.0003 (64-bit)
```

### Operating System Details

```shell
NixOS 24.11 with abraunegg/onedrive: v2.5.3
```

### Client Installation Method

From Source

### OneDrive Account Type

Personal

### What is your OneDrive Application Version

2.5.3

### What is your OneDrive Application Configuration

```shell
sync_dir = "/media/documents/nathan/onedrive_nathan_personal"
skip_file = "/Documents/Big Data*|Family Room|/Documents/DataSets*"
```

### What is your 'curl' version

```shell
curl 8.11.0 (x86_64-pc-linux-gnu) libcurl/8.11.0 OpenSSL/3.3.2 zlib/1.3.1 brotli/1.1.0 zstd/1.5.6 libidn2/2.3.7 libpsl/0.21.5 libssh2/1.11.1 nghttp2/1.64.0
Release-Date: 2024-11-06
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd
```

### Where is your 'sync_dir' located

Local

### What are all your system 'mount points'

```shell
/dev/sdb1 on /media/documents type btrfs (rw,noatime,degraded,compress=lzo,discard=async,space_cache,autodefrag,subvolid=2339,subvol=/documents,x-systemd.mount-timeout=5min)
```

### What are all your local file system partition types

```shell
NAME        FSTYPE FSVER LABEL UUID                                 FSAVAIL FSUSE% MOUNTPOINTS
sda
└─sda1      btrfs        disks 26abb689-0ef7-4034-ab2d-5fc5f6ef3cc4
sdb
└─sdb1      btrfs        disks 26abb689-0ef7-4034-ab2d-5fc5f6ef3cc4    4.4T    52% /media/documents
                                                                                   /media/media
                                                                                   /media/archive
sdc
└─sdc1      btrfs        disks 26abb689-0ef7-4034-ab2d-5fc5f6ef3cc4
sdd
└─sdd1      btrfs        disks 26abb689-0ef7-4034-ab2d-5fc5f6ef3cc4
nvme0n1
├─nvme0n1p1 vfat   FAT32 boot  9FE8-5572                             488.4M     4% /boot
├─nvme0n1p2 ext4   1.0   nixos daff3c15-780e-481d-8a46-4b617b893f00  628.4G    26% /nix/store
│                                                                                  /
└─nvme0n1p3 swap   1     swap  c5c7ac85-74d2-442e-a201-db80f28f5439                [SWAP]
```

### How do you use 'onedrive'

I use a Onedrive together with a couple of Windows machines. There is a big shared folder of Pictures with another Onedrive account.

### Steps to reproduce the behaviour

Have not found it

### Complete Verbose Log Output

```shell
onedrive --confdir=/var/lib/onedrive/onedrive_nathan_personal --resync --sync --verbose
Reading configuration file: /var/lib/onedrive/onedrive_nathan_personal/config
Configuration file successfully loaded
Using 'user' configuration path for application config and state data: /var/lib/onedrive/onedrive_nathan_personal
WARNING: D-Bus message bus daemon is not available; GUI notifications are disabled
Are you sure you wish to proceed with --resync? [Y/N] y

Deleting the saved application sync status ...
Using IPv4 and IPv6 (if configured) for all network operations
Attempting to contact Microsoft OneDrive Login Service
Successfully reached Microsoft OneDrive Login Service
Checking Application Version ...
Attempting to initialise the OneDrive API ...
Configuring Global Azure AD Endpoints
The OneDrive API was initialised successfully
Opening the item database ...
Application Version:  onedrive v2.5.3
Account Type:         personal
Default Drive ID:     REDACTED
Default Root ID:      REDACTED
Remaining Free Space: 551.80 GB (592490738483 bytes)
Sync Engine Initialised with new Onedrive API instance
All application operations will be performed in the configured local 'sync_dir' directory: /media/documents/nathan/onedrive_nathan_personal
Fetching /delta response from the OneDrive API for Drive ID: REDACTED
Processing API Response Bundle: 1 - Quantity of 'changes|items' in this bundle to process: 207
Processing API Response Bundle: 2 - Quantity of 'changes|items' in this bundle to process: 208
...
The usage of --resync will delete your local 'onedrive' client state, thus no record of your current 'sync status' will exist.
This has the potential to overwrite local versions of files with perhaps older versions of documents downloaded from OneDrive, resulting in local data loss.
If in doubt, backup your local data before using --resync

Performing a last examination of the most recent online data within Microsoft OneDrive to complete the reconciliation process
Fetching /delta response from the OneDrive API for Drive ID: REDACTED
Processing API Response Bundle: 1 - Quantity of 'changes|items' in this bundle to process: 7
Finished processing /delta JSON response from the OneDrive API
Processing 6 applicable changes and items received from Microsoft OneDrive
Processing OneDrive JSON item batch [1/1] to ensure consistent local state
Syncing this OneDrive Personal Shared Folder: Camera Roll
Generating a /delta response from the OneDrive API from this Item ID: REDACTED
Adding OneDrive folder details for processing
Adding 41 OneDrive items for processing from the OneDrive Shared Folder
Finished processing self generated /delta JSON response from the OneDrive API
Processing 42 applicable changes and items received from Microsoft OneDrive
Processing OneDrive JSON item batch [1/1] to ensure consistent local state
The following generated a broken database tree query:
Drive ID: REDACTED
Item ID: REDACTED!111
ERROR: A database consistency issue has been caught. A --resync is needed to rebuild the database.
```

**Comment by abraunegg** [comment]
@nathan-gs 
This is potentially a duplicate of #3000 - as per https://github.com/abraunegg/onedrive?tab=readme-ov-file#basic-troubleshooting-steps please can you build from source and determine if the issue still exists.

If the issue *still* exists, please can you generate a verbose debug log as per https://github.com/abraunegg/onedrive/wiki/Generate-debug-log-for-support and email this for analysis.

On a side note:
```
skip_file = "/Documents/Big Data*|Family Room|/Documents/DataSets*"
```

You appear to be using 'skip_file' instead of 'skip_dir' to skip directories ... please correct your application configuration

**Comment by nathan-gs** [comment]
I'm using the lastest commit from master (and fixed the skip_dirs), unfortunately same error:
```
Scanning the local file system '/media/documents/nathan/onedrive_nathan_personal' for new data to upload
Skipping path - excluded by skip_dir config: ./Documents/OneNote Notebooks/My Notebook
Skipping path - excluded by skip_dir config: ./Family Room
Performing a last examination of the most recent online data within Microsoft OneDrive to complete the reconciliation process
Fetching /delta response from the OneDrive API for Drive ID: f1d2221ff510b35
Processing API Response Bundle: 1 - Quantity of 'changes|items' in this bundle to process: 0
Finished processing /delta JSON response from the OneDrive API
No changes or items that can be applied were discovered while processing the data received from Microsoft OneDrive
Syncing this OneDrive Personal Shared Folder: Camera Roll
Generating a /delta response from the OneDrive API from this Item ID: 5BA2395C131E4CC0!431
Adding OneDrive folder details for processing
Adding 41 OneDrive items for processing from the OneDrive Shared Folder
Finished processing self generated /delta JSON response from the OneDrive API
Processing 42 applicable changes and items received from Microsoft OneDrive
Processing OneDrive JSON item batch [1/1] to ensure consistent local state
The following generated a broken database tree query:
Drive ID: REDACTED
Item ID: REDACTED!111
ERROR: A database consistency issue has been caught. A --resync is needed to rebuild the database.
```

**Comment by abraunegg** [comment]
@nathan-gs 
Please can you generate a verbose debug log and email that for analysis

**Comment by nathan-gs** [comment]
Shared it over email, thx!

**Comment by abraunegg** [comment]
@nathan-gs 
Thanks - will review when possible

**Comment by abraunegg** [comment]
@nathan-gs 
Unfortunately the file provided is not a verbose debug log - it is a standard verbose log.

Please can you re-create the verbose debug log and re-send. Please read https://github.com/abraunegg/onedrive/wiki/Generate-debug-log-for-support to assist you.

**Comment by nathan-gs** [comment]
Shared a new log, thx!

**Comment by nathan-gs** [comment]
For now downgraded to v2.4.25 which still works (cleared out the database files first).

**Comment by abraunegg** [comment]
@nathan-gs 

Please can you respond to my email that has information.

**Comment by abraunegg** [comment]
@nathan-gs

Please can you respond to my email that has information.

**Comment by boblemus** [comment]
Verbose debug log sent yesterday as per the instructions (I have the same issue as @nathan-gs )

**Comment by abraunegg** [comment]
@boblemus 
> Verbose debug log sent yesterday as per the instructions (I have the same issue as @nathan-gs )

No email or anything has been received or even in any spam folder. Please re-check / revalidate you have sent to the correct address.

**Comment by abraunegg** [comment]
@boblemus , @nathan-gs

Please can you test the PR below. This is important as I *cannot* reproduce your error.

First install all the require platform dependencies to build the client on your respective platforms. Please read https://github.com/abraunegg/onedrive/blob/master/docs/install.md#building-from-source---high-level-requirements and then follow correctly for your platform.

Once this is done, to clone the PR to resolve your issue, you can use a script like the following:
```
#!/bin/bash

PR=3047

rm -rf ./onedrive-pr${PR}
git clone https://github.com/abraunegg/onedrive.git onedrive-pr${PR}
cd onedrive-pr${PR}
git fetch origin pull/${PR}/head:pr${PR}
git checkout pr${PR}

# Configure and Build
./configure --enable-debug --enable-notifications; make clean; make;
./onedrive --version
```

This script will create a local folder called `onedrive-pr3047` with the PR version.

To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any other options needed>
```
To install the PR, you will need to perform `sudo make install` to install the PR version to your system.

When running the PR, your version should be: `onedrive v2.5.3-27-g627dcfe` or greater.

**Comment by abraunegg** [comment]
@boblemus , @nathan-gs

Please can you respond with your testing of the provided PR to resolve your issue. Please test PR with urgency. I cannot reproduce your issue. This bug is holding up the v2.5.4 release train now.

**Comment by boblemus** [comment]
I can give it a shot tomorrow (in about 36 hours)

**Comment by nathan-gs** [comment]
Also testing it out!

**Comment by nathan-gs** [comment]
Good news @abraunegg, I don't have any errors anymore, sync completed successfully.

**Comment by abraunegg** [comment]
@nathan-gs 
Thanks for the confirmation. Will mark this as 'fixed' and merge into master.

**Comment by nathan-gs** [comment]
Unfortunately I see an error...! The error is gone, but the shared folder causing the issue is also empty.

**Comment by abraunegg** [comment]
> Unfortunately I see an error...! The error is gone, but the shared folder causing the issue is also empty.

Is that shared online folder meant to be empty ? 

Where is it empty ? Online ? Locally ? Please provide more details.

Please remember - I *cannot* reproduce this.

**Comment by nathan-gs** [comment]
Empty locally, full of images remote.

**Comment by abraunegg** [comment]
> Empty locally, full of images remote.

Have you performed a `--resync` ?

As the application was failing (before bug fix), Microsoft only sends certain JSON elements *once* , which is why performing a resync is required to ensure everything is sent for processing. To perform a resync:

```
onedrive --sync --verbose --resync
```

If the local folder is *still* empty post this step, please generate a *new* debug log:
```
onedrive --sync --verbose --verbose --resync --resync-auth > debug_output.log 2>&1
```

**Comment by nathan-gs** [comment]
Ok, will try.

In 2.4.25 it works flawlessly.

**Comment by abraunegg** [comment]
> Ok, will try.
> 
> In 2.4.25 it works flawlessly.

Whilst v2.4.25 works flawlessly - it took ~4 years to iron out *all* the bugs.

v2.5.x is a 100% re-write to support multi-threading. New bugs will exists. v2.5.x went through ~9 months of extensive testing and RC series of releases. I fix the bugs as proactively as I can in an expedient manner to minimise user impact.

I had hoped that folk would test and provide feedback during that process ... obviously not enough people tested and I cannot test every single permutation and implementation people use + the oddities with the European OneDrive Service (it operates 100% different to the rest of the world - and provides JSON responses with missing data)

**Comment by nathan-gs** [comment]
Not a complaint, it is OSS, and your work is very much appreciated!

On the European DC issue, can you send me a detailed message on my email, will see if I can escalate it internally (Microsoft is my employer, but I have nothing to do with Onedrive/Office/...).

**Comment by boblemus** [comment]
Unfortunately I still get the same error. I am currently doing a full
resync (will take a while) and then I'll upload the debug log as per the
instructions.

**Comment by boblemus** [comment]
Link to debug log sent per mail to ***@***.***(redacted)

**Comment by abraunegg** [comment]
@boblemus 
> Unfortunately I still get the same error. I am currently doing a full resync (will take a while) and then I'll upload the debug log as per the instructions.


Unfortunately you are *still* using the v2.5.3 *master* release .. not the PR version as evidenced in your debug log:

```
DEBUG: Application version                          = onedrive v2.5.3-1+np1+1.1
```

#### Steps to resolve
1. Uninstall the *package* 'onedrive' 
2. Compile the application manually as per the instructions above
3. Install the application as per the the instructions above

**Comment by boblemus** [comment]
Sorry, my fault. I have now obtained the PR version and am using this.
Now the resync error is gone but an oddity has occurred - there is a folder
in the root of OneDrive called 2022 which does contain items, both online
and on my Windows PC, but first thing the new onedrive client did was to
delete them locally and since then not sync the files again.

Log files can be downloaded at the same URL as last time.

**Comment by abraunegg** [comment]
@boblemus 
> Sorry, my fault. I have now obtained the PR version and am using this.
Now the resync error is gone but an oddity has occurred - there is a folder
in the root of OneDrive called 2022 which does contain items, both online
and on my Windows PC, but first thing the new onedrive client did was to
delete them locally and since then not sync the files again.

When using the PR version for the first time - did you perform a `--resync`  ?

**Comment by boblemus** [comment]
Yes, I did. Without verbose logging. Once completed I restarted it with all
the parameters (including --resync) as specified in the Wiki for debugging.

**Comment by abraunegg** [comment]
@boblemus 
Looking at the debug log:
```
DEBUG: WARNING: Your cURL/libcurl version (7.81.0) has known HTTP/2 bugs that impact the use of this client.
DEBUG:          Please report this to your distribution, requesting an update to a newer cURL version, or consider upgrading it yourself for optimal stability.
DEBUG:          Downgrading all client operations to use HTTP/1.1 to ensure maximum operational stability.
DEBUG:          Please read https://github.com/abraunegg/onedrive/blob/master/docs/usage.md#compatibility-with-curl for more information.
```

You MUST upgrade your curl version on your system please - it has many errors that cause issues for this client. You are most likely using a Ubuntu LTS platform which stands for Long Term Stuffed .. please upgrade your version of 'curl' or complain loudly to Ubuntu for them to provide a newer version ... or use a much better distribution. To help in updating 'curl' please read: https://github.com/abraunegg/onedrive/blob/master/docs/usage.md#compatibility-with-curl


>  there is a folder
in the root of OneDrive called 2022 which does contain items, both online
and on my Windows PC, but first thing the new onedrive client did was to
delete them locally and since then not sync the files again.

Looking at the PR debug log specifically for this, there are zero deletion events for ./2022 .. and the folder path exists in the log file and has data that it processes - for example:
```
DEBUG: Attempting to calculate local filesystem path for 192*********992 and 1920*********92!***
DEBUG: New Item calculated full path is: 2022/Reg****************teri.pdf
DEBUG: skip_file item to check (full calculated path): /2022/Reg****************teri.pdf
DEBUG: skip_file evaluation for: /2022/Reg****************teri.pdf
DEBUG: Result: false
```

So either:
* This was due to not running the PR version correctly, not doing a --resync
* Still having an older client version running in the background
  * As you are using Ubuntu - if you have *ever* installed the client using the Ubuntu packages, you have a *hidden* systemd service potentially running ... that needs to be disabled and deleted and is generally done manually when you install the client from the OpenSuSE Build Service repository. That is a bug specifically with the Ubuntu packages and not this client.

**Comment by boblemus** [comment]
I started the upgrade process of the LTS version about an hour ago. I'll
let you know how things fare once this is completed.

**Comment by abraunegg** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
